app/tokenMigration.go

```diff
@@ -0,0 +1,139 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math"
+
+	clpkeeper "github.com/Sifchain/sifnode/x/clp/keeper"
+	"github.com/Sifchain/sifnode/x/clp/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/query"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+)
+
+// GetTokenMigrationFunc return a function to migrate token denom
+func GetTokenMigrationFunc(app *SifchainApp) func(ctx sdk.Context, plan upgradetypes.Plan) {
+	return func(ctx sdk.Context, plan upgradetypes.Plan) {
+		ctx.Logger().Info("Starting to execute token migration for balance pool and liquidity")
+
+		ExportAppState("changePoolFormula", app, ctx)
+
+		tokenMap := readTokenMapJSON()
+
+		migrateBalance(ctx, tokenMap, app.BankKeeper)
+		migratePool(ctx, tokenMap, app.ClpKeeper)
+		migrateLiquidity(ctx, tokenMap, app.ClpKeeper)
+	}
+}
+
+func readTokenMapJSON() map[string]string {
+	data, err := ioutil.ReadFile("./token_migration.json")
+	if err != nil {
+		panic(err)
+	}
+
+	jsonData := map[string]string{}
+	err = json.Unmarshal(data, &jsonData)
+	if err != nil {
+		panic("fail to parse token migration file")
+	}
+	return jsonData
+}
+
+func getAll(addresses *[]sdk.AccAddress, coins *[]sdk.Coin) func(address sdk.AccAddress, coin sdk.Coin) bool {
+	return func(address sdk.AccAddress, coin sdk.Coin) bool {
+		*addresses = append(*addresses, address)
+		*coins = append(*coins, coin)
+		return true
+	}
+}
+
+func migrateBalance(ctx sdk.Context, tokenMap map[string]string, bankKeeper bankkeeper.Keeper) {
+	addresses := []sdk.AccAddress{}
+	coins := []sdk.Coin{}
+
+	bankKeeper.IterateAllBalances(ctx, getAll(&addresses, &coins))
+
+	for index, address := range addresses {
+
+		coin := coins[index]
+		amount := coin.Amount
+		// clear the balance for old denom
+		coin.Amount = sdk.NewInt(0)
+		err := bankKeeper.SetBalance(ctx, address, coin)
```

Comment:
Looks like we are zeroing the user's balance here, and updating his balance at L74. Should we move tokenMap check (L72) up so he wont end up in a situation where he's zeroed and the token isnt in the tokenMap?

Reply:
Also, I think we should check if the coins is already a migrated denom, because:
we panic whenever there's an issue, so we might have to rerun migration halfway through. This means there'll be coins already using the new denom. 
In current flow, we'll zero out his balance. Then, in L72, we'll use the migrated denom to lookup tokenMap, which will fail. His balance never gets refilled.